File: Main Source/InputProviders/TouchLib/Source/InputProvider.cs
using System;
using System.AddIn;
using System.Diagnostics;
using System.Threading;
using System.Windows.Forms;
using Multitouch.Contracts;
using TouchlibWrapper;

namespace TouchLib
{
	[AddIn("TouchLib", Publisher = "Daniel Danilin", Description = "Provides input from touchLib (http://nuigroup.com/touchlib)", Version = VERSION)]
	public class InputProvider : IProvider
	{
		CsTI instance;
		int screenWidth;
		int screenHeight;
		internal const string VERSION = "1.0.0.0";

		public event EventHandler<ContactChangedEventArgs> ContactChanged;

		public void Start()
		{
			screenWidth = Screen.PrimaryScreen.Bounds.Width;
			screenHeight = Screen.PrimaryScreen.Bounds.Height;

			instance = CsTI.Instance();
			instance.fingerDown += new fingerDownHandler(instance_fingerDown);
			instance.fingerUp += new fingerUpHandler(instance_fingerUp);
			instance.fingerUpdate += new fingerUpdateHandler(instance_fingerUpdate);


			Thread t = new Thread(ThreadWorker);
			t.SetApartmentState(ApartmentState.MTA);
			t.IsBackground = true;
			t.Name = "touchlib thread";
			t.Start();
		}

		void ThreadWorker()
		{
			if (instance != null)
			{
				IsRunning = true;
				instance.startScreen(true);
			}
		}

		public void Stop()
		{
			if(instance != null)
				instance.stopScreen();
			instance = null;
			IsRunning = false;
		}

		public bool IsRunning { get; private set; }

		private void OnContactChanged(ContactChangedEventArgs e)
		{
			if (ContactChanged != null)
				ContactChanged(this, e);
		}

		void instance_fingerUpdate(int id, int tagID, float x, float y, float angle, float area, float height, float width, float dX, float dY)
		{
			x = x * screenWidth;
			y = y * screenHeight;
			OnContactChanged(new TouchLibContactChangedEventArgs(new TouchLibContact(id, x, y, width, height, ContactState.Moved)));
		}

		void instance_fingerUp(int id, int tagID, float x, float y, float angle, float area, float height, float width, float dX, float dY)
		{
			x = x * screenWidth;
			y = y * screenHeight;
			OnContactChanged(new TouchLibContactChangedEventArgs(new TouchLibContact(id, x, y, width, height, ContactState.Removed)));
		}

		void instance_fingerDown(int id, int tagID, float x, float y, float angle, float area, float height, float width, float dX, float dY)
		{
			x = x * screenWidth;
			y = y * screenHeight;
			OnContactChanged(new TouchLibContactChangedEventArgs(new TouchLibContact(id, x, y, width, height, ContactState.New)));
		}
	}
}

File: Main Source/InputProviders/TouchLibProvider/InputProvider.cpp
#include "StdAfx.h"
#include "InputProvider.h"
#include "TouchListener.h"

#include "TouchLibContactChangedEventArgs.h"

using namespace System::Threading;
using namespace Multitouch::Contracts;

namespace TouchLibProvider
{
	InputProvider::InputProvider(void)
	{}

	void InputProvider::FingerDown(TouchData data)
	{
		Input(this, gcnew TouchLibContactChangedEventArgs(data, ContactState::New));
	}

	void InputProvider::FingerUp(TouchData data)
	{
		Input(this, gcnew TouchLibContactChangedEventArgs(data, ContactState::Removed));
	}

	void InputProvider::FingerUpdate(TouchData data)
	{
		Input(this, gcnew TouchLibContactChangedEventArgs(data, ContactState::Moved));
	}

	bool InputProvider::IsRunning::get(void)
	{
		return isRunning;
	}

	void InputProvider::Start(void)
	{
		if(listener)
			Stop();

		screenBounds = System::Windows::Forms::Screen::PrimaryScreen->Bounds;

		listener = new TouchListener(this, true);
		Thread^ thread = gcnew Thread(gcnew ThreadStart(this, &TouchLibProvider::InputProvider::OnThreadStart));
		thread->Start();
	}

	void InputProvider::OnThreadStart(void)
	{
		listener->Start();
		isRunning = true;
	}

	void InputProvider::Stop(void)
	{
		isRunning = false;
		listener->Stop();
		delete listener;
	}

	System::Drawing::Rectangle InputProvider::ScreenBounds::get(void)
	{
		return screenBounds;
	}
}